codec/src/main/java/io/netty/handler/codec/compression/ZstdDecoder.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright 2023 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.compression;
+
+import com.github.luben.zstd.ZstdInputStreamNoFinalizer;
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.ByteToMessageDecoder;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.List;
+
+/**
+ * Decompresses a compressed block {@link ByteBuf} using the Zstandard algorithm.
+ * See <a href="https://facebook.github.io/zstd">Zstandard</a>.
+ */
+public final class ZstdDecoder extends ByteToMessageDecoder {
+    private final MutableByteBufInputStream inputStream = new MutableByteBufInputStream();
+    private ZstdInputStreamNoFinalizer zstdIs;
+
+    private State currentState = State.DECOMPRESS_DATA;
+
+    /**
+     * Current state of stream.
+     */
+    private enum State {
+        DECOMPRESS_DATA,
+        CORRUPTED
+    }
+
+    @Override
+    protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
+        try {
+            if (currentState == State.CORRUPTED) {
+                in.skipBytes(in.readableBytes());
+                return;
+            }
+            final int compressedLength = in.readableBytes();
+
+            inputStream.current = in;
+
+            ByteBuf outBuffer = null;
+            try {
+                int w = -1;
+                do {
+                    if (outBuffer == null) {
+                        // Let's start with the compressedLength * 2 as often we will not have everything
+                        // we need in the in buffer and don't want to reserve too much memory.
+                        outBuffer = ctx.alloc().heapBuffer(compressedLength * 2);
+                    }
+                    byte[] array = outBuffer.array();
+                    int writerOffset = outBuffer.arrayOffset() + outBuffer.writerIndex();
+                    int writableBytes = outBuffer.writableBytes();
+                    int written = 0;
+                    while (writableBytes > 0 && (w = zstdIs.read(array, writerOffset, writableBytes)) != -1) {
+                        writerOffset += w;
+                        writableBytes -= w;
+                        written += w;
+                    }
+                    outBuffer.writerIndex(outBuffer.writerIndex() + written);
```

Comment:
This can also be written like this:
```suggestion
                    do {
                        w = outBuffer.writeBytes(zstdIs, outBuffer.writableBytes());
                    } while (w != -1 && outBuffer.isWritable());
```